Add secrets backend preview deployment info (#3249)
astronomer/docs: astro/ci-cd-templates/github-actions.md

The Astronomer [Deploy Action](https://github.com/astronomer/deploy-action/tree/deployment-preview#deployment-preview-templates) includes several sub-actions that can be used together to create a complete [Deployment preview](ci-cd-templates/template-overview.md#preview-deployment-templates) pipeline.

> You can choose to save your access credential information in either Github Secrets or with your own secrets backend. If you use a [secrets backend](secrets-backend.md) to manage your credentials, you can configure your Github Action to use and store them.

Reviewer: From what I can tell from the feature and templates, the purpose of the secrets backend is not to access the same type of credentials that you would access from GitHub secrets. GH secrets stores credentials related to CI/CD itself, while the secrets backend stores information that your Deployment needs to run work, such as connection information, DAG-level variables, etc.

Maybe we can shift the framing away from just "credentials" and instead refer to "Airflow objects and credentials stored in your secrets backend." Our secrets backend docs are framed primarily about configuring backends to store Airflow connections and variables

> ### Standard Deployment preview

Reviewer: ```suggestion
### Standard Deployment preview template
```

The Astronomer [Deploy Action](https://github.com/astronomer/deploy-action/tree/deployment-preview#deployment-preview-templates) includes several sub-actions that can be used together to create a complete [Deployment preview](ci-cd-templates/template-overview.md#preview-deployment-templates) pipeline.

> You can choose to use

Reviewer 2: ?

            deployment-id: <main-deployment-id>
    ```

4. (Optional) If you want to use a secrets backend instead of GitHub secrets, add the following code to the end of the `steps` section in your `create-deployment-preview.yml` file.

  ```yaml
            - name: Create Secret Variables
            run: |

>               astro deployment variable update --deployment-id ${{steps.create-dep-prev.outputs.preview-id}} AIRFLOW__SECRETS__BACKEND_KWARGS=${{ secrets.AIRFLOW__SECRETS__BACKEND_KWARGS }} --secret

Reviewer 3: `steps.create-dep-prev.outputs.preview-id` does not work unless you have a step with id set to `create-dep-prev`. For example:

```
steps:
    - name: Create Deployment Preview
      uses: astronomer/deploy-action@v0.3
      with:
        action: create-deployment-preview
        deployment-name: "test"
      id: create-dep-prev 
```

            deployment-id: <main-deployment-id>
    ```

7. In the same folder, create a new YAML file named `deploy-to-main-deployment.yml` that includes the following configuration:

    ```yaml
    name: Astronomer CI - Deploy code to main Deployment

    on:
      push:
        branches:
          - main

    env:
      ## Set your API token as a GitHub secret
      ASTRO_API_TOKEN: ${{ secrets.ASTRO_API_TOKEN }}

    jobs:
      deploy:
        runs-on: ubuntu-latest
        steps:
        - name: Deploy code to main Deployment
          uses: astronomer/deploy-action@v0.2
          with:
            deployment-id: <main-deployment-id>
    ```

    All four workflow files must have the same Deployment ID specified. The actions use this Deployment ID to create and delete preview Deployments based on your main Deployment.

### Deployment preview and secrets backend implementation

#### Prerequisites

- An Astro project hosted in a GitHub repository.
- A [Workspace API token](workspace-api-tokens.md).
- A [Deployment](create-deployment.md).
- A [Secrets backend](secrets-backend.md).

#### Setup

1. Copy and save the Deployment ID for your Astro deployment.

> 2. Set the following [GitHub secrets](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) in the repository hosting your Astro project. This includes your Astro API Token, so that GitHub has permissions to deploy code to your Deployments or Workspaces, and your `AIRFLOW__SECRETS__BACKEND_KWARGS`. You can find these values in your Astro project's `.env`file, replacing `<your-token>` and `<your-kwargs>` with your unique values.

Reviewer 3: I'm not sure that these values would always be in your projects `.env` file. You get the token value when you create an API token and there are many places a user could store this. Same thing with the kwargs. You get them while you are setting up a secrets backend. For these they usually saved as an env var in some way but that doesn't have to be in the `.env` file

Reviewer: Agreed, I don't think we should assume that the user configuring CI/CD has set up secrets backend at the local level first. It's possible that an admin user is configuring this pipeline with credentials that the DAG author doesn't have access to, so we should assume the use has a fresh slate going into this setup.

  - Key: `ASTRO_API_TOKEN`
  - Secret: `<your-token>`

  - Key: `AIRFLOW__SECRETS__BACKEND_KWARGS`
  - Secret: `<your-kwargs>`

  Refer to the [secrets backend](secrets-backend.md) overview and procedures docs to learn more about setting up a secrets backend and defining your `AIRFLOW__SECRETS__BACKEND_KWARGS`.

3. In your project repository, create a new YAML file in `.github/workflows` named `create-deployment-preview.yml` that includes the following configuration.

    ```yaml
    name: Astronomer CI - Create preview Deployment

    on:
      create:
        branches:
          - "**"

    env:
      ## Sets Deployment API key credentials as environment variables
      ASTRO_API_TOKEN: ${{ secrets.ASTRO_API_TOKEN }}

    jobs:
      deploy:
        runs-on: ubuntu-latest
        steps:
        - name: Create preview Deployment
          uses: astronomer/deploy-action@v0.2
          with:
            action: create-deployment-preview

>             deployment-id: <main-deployment-id>

Reviewer 3: ```suggestion
            deployment-id: <main-deployment-id>
          id: create-dep-prev
```
This step needs the id defined to be used in later steps

    All four workflow files must have the same Deployment ID specified. The actions use this Deployment ID to create and delete preview Deployments based on your main Deployment.

### Deployment preview template with secrets backend implementation

> If you use a [secrets backend](secrets-backend.md) to manage your access credentials and other sensitive information, you can configure your Github Action to grant preview Deployments access to your secrets backend. This means that DAGs in the preview Deployment can access secrets such as Airflow variables and connections for testing purposes.

Reviewer: ```suggestion
If you use a [secrets backend](secrets-backend.md) to manage Airflow objects such as variables and connections, you can configure your action to grant preview Deployments access to your secrets backend. This means that DAGs in the preview Deployment can access your secret Airflow objects for testing purposes.
```

> Unlike the standard Deployment preview setup, you need to add some additional configuration to GitHub including information. Specifically, this template makes use of the `AIRFLOW__SECRETS__BACKEND_KWARGS` environment variable to store information and credentials for your secrets backend.

Reviewer: ```suggestion
This template makes use of the `AIRFLOW__SECRETS__BACKEND_KWARGS` environment variable to store information and credentials for your secrets backend.
```

#### Prerequisites

- An Astro project hosted in a GitHub repository.
- A [Workspace API token](workspace-api-tokens.md).
- A [Deployment](create-deployment.md).

> - A [Secrets backend](secrets-backend.md).

Reviewer: ```suggestion
- A [secrets backend](secrets-backend.md).
```

#### Setup

1. Copy and save the Deployment ID for your Astro deployment.
2. Set the following [GitHub secrets](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) in the repository hosting your Astro project. This includes your Astro API Token, so that GitHub has permissions to deploy code to your Deployments or Workspaces, and your secrets backend information stored in `AIRFLOW__SECRETS__BACKEND_KWARGS`. See [Configure a secrets backend](secrets-backend.md) for more information about configuring your secrets backend as an environment variable.

>   - Key 1: `ASTRO_API_TOKEN`
>   - Secret 1: `<your-token>`
>   - Key 2: `AIRFLOW__SECRETS__BACKEND_KWARGS`
>   - Secret 2: `<your-kwargs>`

Reviewer: ```suggestion
    - **Key 1**: `ASTRO_API_TOKEN`
    - **Secret 1**: `<your-token>`
    - **Key 2**: `AIRFLOW__SECRETS__BACKEND_KWARGS`
    - **Secret 2**: `<your-kwargs>`
```

3. In your project repository, create a new YAML file in `.github/workflows` named `create-deployment-preview.yml` that includes the following configuration.

    ```yaml
    name: Astronomer CI - Create preview Deployment with Secrets Backend

    on:
      create:
        branches:
          - "**"

    env:
      ## Sets Deployment API key credentials as environment variables
      ASTRO_API_TOKEN: ${{ secrets.ASTRO_API_TOKEN }}

    jobs:
      deploy:
        runs-on: ubuntu-latest
        steps:
        - steps:
          - name: Create Deployment Preview
            uses: astronomer/deploy-action@v0.3
            with:
              action: create-deployment-preview
              deployment-name: "test"
              id: create-dep-prev
          - name: Create Secret Variables
            run: |
              astro deployment variable update --deployment-id ${{steps.create-dep-prev.outputs.preview-id}} AIRFLOW__SECRETS__BACKEND_KWARGS=${{ secrets.AIRFLOW__SECRETS__BACKEND_KWARGS }} --secret

>   ```

Reviewer: ````suggestion
    ```
````

Reviewer: This should fix the error I'm seeing in preview, but I'd double check just in case it's something else as well

    All four workflow files must have the same Deployment ID specified. The actions use this Deployment ID to create and delete preview Deployments based on your main Deployment.

> ### Secrets backend Deployment preview

Reviewer: ```suggestion
### Deployment preview template with secrets backend implementation
```

Reviewer: It's wordy, but I also don't think a "Secrets backend preview" is a framing that users will be familiar with.

> If you use a [secrets backend](secrets-backend.md) to manage your access credentials and other sensitive information, you can configure your Github Action to use and store them while working with Deployment previews.

Reviewer: ```suggestion
If you use a [secrets backend](secrets-backend.md) to manage your access credentials and other sensitive information, you can configure your Github Action to grant preview Deployments access to your secrets backend. This means that DAGs in the preview Deployment can access secrets such as Airflow variables and connections for testing purposes.
```

> Unlike the standard Deployment preview setup, you need to add some additional configuration information to both your GitHub Secrets and GitHub workflow include information about your Secrets Backend kwargs. This information instructs the CI/CD pipeline to refer to your Secrets Backend to read and write access credentials.

Reviewer: ```suggestion
Unlike the standard Deployment preview setup, you need to add some additional configuration to GitHub including information. Specifically, this template makes use of the `AIRFLOW__SECRETS__BACKEND_KWARGS` environment variable to store information and credentials for your secrets backend.
```

#### Prerequisites

- An Astro project hosted in a GitHub repository.
- A [Workspace API token](workspace-api-tokens.md).
- A [Deployment](create-deployment.md).
- A [Secrets backend](secrets-backend.md).

#### Setup

1. Copy and save the Deployment ID for your Astro deployment.

> 2. Set the following [GitHub secrets](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) in the repository hosting your Astro project. This includes your Astro API Token, so that GitHub has permissions to deploy code to your Deployments or Workspaces, and your `AIRFLOW__SECRETS__BACKEND_KWARGS`. You can find these values in your Astro project's `.env`file, replacing `<your-token>` and `<your-kwargs>` with your unique values.

Reviewer: ```suggestion
2. Set the following [GitHub secrets](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) in the repository hosting your Astro project. This includes your Astro API Token, so that GitHub has permissions to deploy code to your Deployments or Workspaces, and your secrets backend information stored in `AIRFLOW__SECRETS__BACKEND_KWARGS`. See [Configure a secrets backend](secrets-backend.md) for more information about configuring your secrets backend as an environment variable.
```

>   - Key: `ASTRO_API_TOKEN`
>   - Secret: `<your-token>`
>   - Key: `AIRFLOW__SECRETS__BACKEND_KWARGS`
>   - Secret: `<your-kwargs>`

Reviewer: ```suggestion
  - Key 1: `ASTRO_API_TOKEN`
  - Secret 1: `<your-token>`
  - Key 2: `AIRFLOW__SECRETS__BACKEND_KWARGS`
  - Secret 2: `<your-kwargs>`
```

> 
>   Refer to the [secrets backend](secrets-backend.md) overview and procedures docs to learn more about setting up a secrets backend and defining your `AIRFLOW__SECRETS__BACKEND_KWARGS`.

Reviewer: ```suggestion
```
